src/pykeen/models/nbase.py

```diff
@@ -427,94 +313,96 @@ def append_weight_regularizer(
+            return scores
+        return scores.repeat(1, self.num_entities)
 
-    def _repeat_scores_if_necessary(
-        self,
-        scores: torch.FloatTensor,
-        h_indices: Optional[torch.LongTensor],
-        r_indices: Optional[torch.LongTensor],
-        t_indices: Optional[torch.LongTensor],
-    ) -> torch.FloatTensor:
-        """Repeat scores for entities/relations if the model does not have representations for one of them."""
-        repeat_relations = len(self.relation_representations) == 0
-        repeat_entities = len(self.entity_representations) == 0
-
-        if not (repeat_entities or repeat_relations):
+    def score_h(self, rt_batch: torch.LongTensor, slice_size: Optional[int] = None) -> torch.FloatTensor:
+        """Forward pass using left side (head) prediction.
+
+        This method calculates the score for all possible heads for each (relation, tail) pair.
+
+        :param rt_batch: shape: (batch_size, 2), dtype: long
+            The indices of (relation, tail) pairs.
+        :param slice_size:
+            The slice size.
+
+        :return: shape: (batch_size, num_entities), dtype: float
+            For each r-t pair, the scores for all possible heads.
+        """
+        h, r, t = self._get_representations(h=None, r=rt_batch[:, 0], t=rt_batch[:, 1])
+        scores = self.interaction.score_h(all_entities=h, r=r, t=t, slice_size=slice_size)
+        if self.entity_representations:
             return scores
+        return scores.repeat(1, self.num_entities)
 
-        repeats = [1, 1, 1, 1]
+    def score_r(self, ht_batch: torch.LongTensor, slice_size: Optional[int] = None) -> torch.FloatTensor:
+        """Forward pass using middle (relation) prediction.
 
-        for i, (flag, ind, num) in enumerate(
-            (
-                (repeat_entities, h_indices, self.num_entities),
-                (repeat_relations, r_indices, self.num_relations),
-                (repeat_entities, t_indices, self.num_entities),
-            ),
-            start=1,
-        ):
-            if flag:
-                if ind is None:
-                    repeats[i] = num
-                else:
-                    batch_size = len(ind)
-                    if scores.shape[0] < batch_size:
-                        repeats[0] = batch_size
-
-        return scores.repeat(*repeats)
+        This method calculates the score for all possible relations for each (head, tail) pair.
+
+        :param ht_batch: shape: (batch_size, 2), dtype: long
+            The indices of (head, tail) pairs.
+        :param slice_size:
+            The slice size.
+
+        :return: shape: (batch_size, num_relations), dtype: float
+            For each h-t pair, the scores for all possible relations.
+        """
+        h, r, t = self._get_representations(h=ht_batch[:, 0], r=None, t=ht_batch[:, 1])
+        scores = self.interaction.score_r(h=h, all_relations=r, t=t, slice_size=slice_size)
+        if self.relation_representations:
```

Comment:
can you add a comment why you're doing this conditional

Reply:
d87c20e